modulemd/modulemd-module-stream.c

```diff
@@ -259,18 +262,35 @@ modulemd_module_stream_read_yaml (yaml_parser_t *parser,
     case MD_MODULESTREAM_VERSION_THREE:
       if (doctype == MODULEMD_YAML_DOC_PACKAGER)
         {
-          g_set_error (error,
-                       MODULEMD_YAML_ERROR,
-                       MMD_YAML_ERROR_PROGRAMMING,
-                       "Incorrect function to parse modulemd-packager v3");
-          return NULL;
+          packager_v3 =
+            modulemd_packager_v3_parse_yaml (subdoc, &nested_error);
+          if (!packager_v3)
+            {
+              g_propagate_error (error, g_steal_pointer (&nested_error));
+              return NULL;
+            }
+
+          /* convert packager v3 to stream v2 for backwards compatibility */
+          stream = MODULEMD_MODULE_STREAM (
+            modulemd_packager_v3_to_stream_v2 (packager_v3, &nested_error));
+          if (!stream)
+            {
+              g_propagate_error (error, g_steal_pointer (&nested_error));
+              return NULL;
+            }
```

Comment:
We need to conditionalize this, because if we are expecting to be v3, we don't want to downgrade and upgrade later. (Among other things, this would result in losing the per-configuration buildopts).

Reply:
I thought we discussed doing this was the key to provided backwards compatibility to existing consumers of libmodulemd. (eg., RHEL 8) Do you have any suggestions for how the API should be modified/extended so the preference for StreamV2 vs StreamV3 can be indicated?

Reply:
The conversion from PackagerV3->StreamV2 is slightly lossy due to the buildopts. If we don't want to lose that, we'll need to retain that data to re-add it for a later upgrade.

Maybe we could have `modulemd_packager_v3_to_stream_v2()` store a copy of the original `PackagerV3` object internally and in the event that the consumer calls `modulemd_modulestream_upgrade()`, we produce the results converted from the Packager object instead of the StreamV2 object? We'd have to make sure they don't get out of sync; maybe we can have it so that the stored Packager object is freed if any "write" to the StreamV2 object occurs?

Reply:
I implemented this per our earlier discussion by creating a new configuration setting and API for specifying the desired default stream metadata version, and using that setting to determine if the packager v3 document gets returned as a StreamV2 or StreamV3 object.

Note: since this function returns a single Stream, it will fail if the packager v3 document contains multiple build configurations--which causes it to expand to multiple StreamV3s.